Implement new Transaction Broadcast Protocol using base node RPC service

philipr-za opened the pull request:
## Description

*The PR is quite large but a lot of it is updating the existing Transaction Service Tests. I will mark the specific files that should be reviewed.*

This PR implements a new Transaction Broadcast Protocol that makes use of the new Base Node RPC service built for the wallet . This removes the use P2P fire-and-forget communication stack for broadcasting and querying transactions in the base node. This greatly simplifies the process of communicating with the base node as all the edge cases produced by using fire and forget messaging are removed.

The new protocol will submit a transaction to the base node to be included in the mempool. The base node will now immediately reject a transaction if it is time locked, an orphan, a double spend , already mined or otherwise fails validation. In the past the transaction would end up in the orphan, pending or reorg pool instead of being immediately rejected. 

Once the transaction is in the mempool the wallet will then poll the base node to find out if the transaction is mined, but in an unconfirmed state (i.e. has less that the threshold confirmations) and keep polling until the transaction has received enough confirmations to be considered mined. The addition of monitoring confirmations is a new feature that the base node provides and this new protocol makes use of.



## How Has This Been Tested?
The new protocol is extensively tested as a module by using a mock RPC server to stand in for a real base node. The existing transaction service tests are updated to reflect the new architecture and many of the long-winded tests that used to test the base node broadcast behaviour using the fire-and-forget messaging have been removed in favour of these new tests.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I ran `cargo test` successfully before submitting my PR.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

philipr-za commented on base_layer/wallet/src/transaction_service/protocols/transaction_broadcast_protocol.rs:
```diff
@@ -20,486 +20,508 @@
 where TBackend: TransactionBackend + 'static
 {
-    id: u64,
+    tx_id: TxId,
```
👀  on this whole module

philipr-za commented on base_layer/wallet/src/transaction_service/service.rs:
```diff
@@ -55,18 +54,12 @@ use futures::{
-    time::Duration,
-};
-use tari_comms::{peer_manager::NodeIdentity, types::CommsPublicKey};
+use std::{collections::HashMap, convert::TryInto, sync::Arc, time::Duration};
```
👀

philipr-za commented on base_layer/wallet/tests/support/rpc.rs:
```diff
@@ -0,0 +1,266 @@
+}
+
+#[derive(Clone)]
+pub struct BaseNodeWalletRpcMockState {
```
👀 
@delta1 this module might be of interest on the work you are currently doing as it shows how to set up a mock RPC Server to test you RPC client calls against. Tests at the bottom show how it is used.

delta1 replied:
Brilliant thanks @philipr-za will take a look and also review ASAP

stringhandler commented on base_layer/core/tests/wallet.rs:
```diff
@@ -1,382 +0,0 @@
-}
-static EMISSION: [u64; 2] = [10, 10];
-#[test]
-fn wallet_base_node_integration_test() {
```
Is this moved somewhere else?

philipr-za replied:
This is to become a cucumber test

stringhandler commented on base_layer/wallet/src/transaction_service/protocols/transaction_broadcast_protocol.rs:
```diff
@@ -20,486 +20,508 @@
+                                connection = Some(base_node_connection);
+                            },
+                            Err(e) => {
+                                info!(target: LOG_TARGET, "Problem connecting to base node: {} for Transaction Broadcast Protocol (TxId: {})", e, self.tx_id);
```
Warn perhaps? Makes grepping easier

philipr-za committed:
Implement new Transaction Broadcast Protocol using base node RPC service

This PR implements a new Transaction Broadcast Protocol that makes use of the new Base Node RPC service built for the wallet . This removes the use P2P fire-and-forget communication stack for broadcasting and querying transactions in the base node. This greatly simplifies the process of communicating with the base node as all the edge cases produced by using fire and forget messaging are removed.

The new protocol will submit a transaction to the base node to be included in the mempool. The base node will now immediately reject a transaction if it is time locked, an orphan, a double spend , already mined or otherwise fails validation. In the past the transaction would end up in the orphan, pending or reorg pool instead of being immediately rejected. 

Once the transaction is in the mempool the wallet will then poll the base node to find out if the transaction is mined, but in an unconfirmed state (i.e. has less that the threshold confirmations) and keep polling until the transaction has received enough confirmations to be considered mined. The addition of monitoring confirmations is a new feature that the base node provides and this new protocol makes use of.

# Testing
The new protocol is extensively tested as a module by using a mock RPC server to stand in for a real base node. The existing transaction service tests are updated to reflect the new architecture and many of the long-winded tests that used to test the base node broadcast behaviour using the fire-and-forget messaging have been removed in favour of these new tests.
